File: demo/java/src/main/java/info/chenliang/tank/Item.java
package info.chenliang.tank;
import java.io.*;
import info.chenliang.talky.*;
public class Item{
public byte id;
public String name;
public void serialize(DataOutputStream dos) throws Exception{
FieldMark fm = new FieldMark(1);
fm.mark(id != 0);
fm.mark(name != null && name.length() > 0);
dos.write(fm.getData());
if(fm.isMarked(0)){
dos.writeByte(id);
}
if(fm.isMarked(1)){
dos.writeUTF(name);
}
}
public void deserialize(DataInputStream dis) throws Exception{
FieldMark fm = new FieldMark(1);
dis.read(fm.getData());
if(fm.readMark()){
id = dis.readByte();
}
if(fm.readMark()){
name = dis.readUTF();
}
}
}

File: demo/java/src/main/java/info/chenliang/tank/Item.cs
using System;
using System.IO;
using MiscUtil;
using MiscUtil.IO;
using MiscUtil.Conversion;
using Talky;
namespace info.chenliang.tank{
public class Item{
public sbyte id;
public string name;
public void Serialize(MemoryEndianBinaryWriter writer){
FieldMark fm = new FieldMark(1);
fm.Mark(id != 0);
fm.Mark(name != null && name.Length() > 0);
writer.Write(fm.getData());
if(fm.IsMarked(0)){
writer.WriteInt8(id);
}
if(fm.IsMarked(1)){
writer.WriteUTF(name);
}
}
public void Deserialize(MemoryEndianBinaryReader reader){
FieldMark fm = new FieldMark(1);
reader.Read(fm.getData());
if(fm.ReadMark()){
id = reader.ReadByte();
}
if(fm.ReadMark()){
name = reader.ReadUTF();
}
}
}
}
